src/main/java/com/google/devtools/build/lib/bazel/bzlmod/SingleExtensionEvalFunction.java

```diff
@@ -478,127 +454,429 @@ private BzlLoadValue loadBzlFile(
-      thread.setPrintHandler(Event.makeDebugPrintHandler(env.getListener()));
-      moduleContext = createContext(env, usagesValue, starlarkSemantics, extensionId, extension);
-      threadContext.storeInThread(thread);
-      try (SilentCloseable c =
-          Profiler.instance()
-              .profile(
-                  ProfilerTask.BZLMOD,
-                  () -> "evaluate module extension: " + extensionId.asTargetString())) {
-        Object returnValue =
-            Starlark.fastcall(
-                thread, extension.getImplementation(), new Object[] {moduleContext}, new Object[0]);
-        if (returnValue != Starlark.NONE && !(returnValue instanceof ModuleExtensionMetadata)) {
+      throws InterruptedException, SingleExtensionEvalFunctionException {
+    // An innate extension should have a singular usage.
+    if (usagesValue.getExtensionUsages().size() > 1) {
+      throw new SingleExtensionEvalFunctionException(
+          ExternalDepsException.withMessage(
+              Code.BAD_MODULE,
+              "innate module extension %s is used by multiple modules: %s",
+              extensionId,
+              usagesValue.getExtensionUsages().keySet()),
+          Transience.PERSISTENT);
+    }
+    ModuleKey moduleKey = Iterables.getOnlyElement(usagesValue.getExtensionUsages().keySet());
+    Preconditions.checkState(moduleKey.moduleFileLabel().equals(extensionId.getBzlFileLabel()));
+    ImmutableList<Tag> tags =
+        Iterables.getOnlyElement(usagesValue.getExtensionUsages().values()).getTags();
+    RepositoryMapping repoMapping = usagesValue.getRepoMappings().get(moduleKey);
+
+    // Each tag of this usage defines a repo. The name of the tag is of the form
+    // "<bzl_file_label>%<rule_name>". Collect the .bzl files referenced and load them.
+    Label.RepoContext repoContext =
+        Label.RepoContext.of(moduleKey.getCanonicalRepoName(), repoMapping);
+    ArrayList<Label> loadLabels = new ArrayList<>(tags.size());
+    ArrayList<String> ruleNames = new ArrayList<>(tags.size());
+    for (Tag tag : tags) {
+      Iterator<String> parts = Splitter.on('%').split(tag.getTagName()).iterator();
+      try {
+        Label label = Label.parseWithRepoContext(parts.next(), repoContext);
+        BzlLoadFunction.checkValidLoadLabel(label, starlarkSemantics);
+        loadLabels.add(label);
+        ruleNames.add(parts.next());
+      } catch (LabelSyntaxException e) {
+        throw new SingleExtensionEvalFunctionException(
+            ExternalDepsException.withCauseAndMessage(
+                Code.BAD_MODULE, e, "bad repo rule .bzl file label at %s", tag.getLocation()),
+            Transience.PERSISTENT);
+      }
+    }
+    ImmutableSet<BzlLoadValue.Key> loadKeys =
+        loadLabels.stream().map(BzlLoadValue::keyForBzlmod).collect(toImmutableSet());
+    SkyframeLookupResult loadResult = env.getValuesAndExceptions(loadKeys);
+    LinkedHashMap<Label, BzlLoadValue> loadedBzls = new LinkedHashMap<>();
```

Comment:
Could be a pre-sized `ImmutableMap.Builder`.

Reply:
obsolete. But just for the record, LinkedHashMap works better than an ImmutableMap.Buidler in the case where there are many duplicate keys, since the latter stores all entries in an array before the build() call. LinkedHashMap could save memory in that case.